## Assignment Name: Pandas 1 - **Data Analysis on Weather Data**

**Problem Statement**

  Assume that you are working as a Data Analyst Intern with IMD - India Meteorological Department.
  Your first assignment as an intern here is to perform analysis on weather data recorded between 1st Dec 2008 and 25th June 2017 in Australia.

Based on the given dataset we have to predict wheater RainTomorrow  or not

    Date - The date of observation
    Location - The common name of the location of the weather station
    MinTemp - The minimum temperature in degrees celsius
    MaxTemp - The maximum temperature in degrees celsius
    Rainfall - The amount of rainfall recorded for the day in mm
    Evaporation - The so-called Class A pan evaporation (mm) in the 24 hours to 9am
    Sunshine - The number of hours of bright sunshine in the day.
    WindGustDir - The direction of the strongest wind gust in the 24 hours to midnight
    WindGustSpeed - The speed (km/h) of the strongest wind gust in the 24 hours to midnight
    WindDir9am - Direction of the wind at 9am
    WindDir3pm - Direction of the wind at 3pm
    WindSpeed9am - Wind speed (km/hr) averaged over 10 minutes prior to 9am
    WindSpeed3pm - Wind speed (km/hr) averaged over 10 minutes prior to 3pm
    Humidity9am - Humidity (percent) at 9am
    Humidity3pm - Humidity (percent) at 3pm
    Pressure9am - Atmospheric pressure (hpa) reduced to mean sea level at 9am
    Pressure3pm - Atmospheric pressure (hpa) reduced to mean sea level at 3pm
    Cloud9am - Fraction of sky obscured by cloud at 9am. This is measured in "oktas", which are a unit of eigths. It records how many eigths of the sky are obscured by cloud. A 0 measure indicates completely clear sky whilst an 8 indicates that it is completely overcast.
    Cloud3pm - Fraction of sky obscured by cloud (in "oktas": eighths) at 3pm. See Cload9am for a description of the values
    Temp9am - Temperature (degrees C) at 9am
    Temp3pm - Temperature (degrees C) at 3pm
    RainToday - Boolean: 1 if precipitation (mm) in the 24 hours to 9am exceeds 1mm, otherwise 0
    RainTomorrow - The amount of next day rain in mm. Used to create response variable RainTomorrow. A kind of measure of the "risk".

In [ ]:
#importing libraries
import numpy as np
import pandas as pd

In [ ]:
df=pd.read_csv("/content/weather_data.csv")

In [ ]:
df

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0       12/1/2008   Albury     13.4     22.9       0.6          NaN       NaN   
1       12/2/2008   Albury      7.4     25.1       0.0          NaN       NaN   
2       12/3/2008   Albury     12.9     25.7       0.0          NaN       NaN   
3       12/4/2008   Albury      9.2     28.0       0.0          NaN       NaN   
4       12/5/2008   Albury     17.5     32.3       1.0          NaN       NaN   
...           ...      ...      ...      ...       ...          ...       ...   
145455  6/21/2017    Uluru      2.8     23.4       0.0          NaN       NaN   
145456  6/22/2017    Uluru      3.6     25.3       0.0          NaN       NaN   
145457  6/23/2017    Uluru      5.4     26.9       0.0          NaN       NaN   
145458  6/24/2017    Uluru      7.8     27.0       0.0          NaN       NaN   
145459  6/25/2017    Uluru     14.9      NaN       0.0          NaN       NaN   

       WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0                W           44.0          W  ...        71.0         22.0   
1              WNW           44.0        NNW  ...        44.0         25.0   
2              WSW           46.0          W  ...        38.0         30.0   
3               NE           24.0         SE  ...        45.0         16.0   
4                W           41.0        ENE  ...        82.0         33.0   
...            ...            ...        ...  ...         ...          ...   
145455           E           31.0         SE  ...        51.0         24.0   
145456         NNW           22.0         SE  ...        56.0         21.0   
145457           N           37.0         SE  ...        53.0         24.0   
145458          SE           28.0        SSE  ...        51.0         24.0   
145459         NaN            NaN        ESE  ...        62.0         36.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
1            1010.6       1007.8       NaN       NaN     17.2     24.3   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
3            1017.6       1012.8       NaN       NaN     18.1     26.5   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
...             ...          ...       ...       ...      ...      ...   
145455       1024.6       1020.3       NaN       NaN     10.1     22.4   
145456       1023.5       1019.1       NaN       NaN     10.9     24.5   
145457       1021.0       1016.8       NaN       NaN     12.5     26.1   
145458       1019.4       1016.5       3.0       2.0     15.1     26.0   
145459       1020.2       1017.9       8.0       8.0     15.0     20.9   

        RainToday  RainTomorrow  
0              No            No  
1              No            No  
2              No            No  
3              No            No  
4              No            No  
...           ...           ...  
145455         No            No  
145456         No            No  
145457         No            No  
145458         No            No  
145459         No           NaN  

[145460 rows x 23 columns]

In [ ]:
df.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  12/1/2008   Albury     13.4     22.9       0.6          NaN       NaN   
1  12/2/2008   Albury      7.4     25.1       0.0          NaN       NaN   
2  12/3/2008   Albury     12.9     25.7       0.0          NaN       NaN   
3  12/4/2008   Albury      9.2     28.0       0.0          NaN       NaN   
4  12/5/2008   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

With the df.head() we can see the top 5 rows and 23 columns

In [ ]:
df.tail()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
145455  6/21/2017    Uluru      2.8     23.4       0.0          NaN       NaN   
145456  6/22/2017    Uluru      3.6     25.3       0.0          NaN       NaN   
145457  6/23/2017    Uluru      5.4     26.9       0.0          NaN       NaN   
145458  6/24/2017    Uluru      7.8     27.0       0.0          NaN       NaN   
145459  6/25/2017    Uluru     14.9      NaN       0.0          NaN       NaN   

       WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
145455           E           31.0         SE  ...        51.0         24.0   
145456         NNW           22.0         SE  ...        56.0         21.0   
145457           N           37.0         SE  ...        53.0         24.0   
145458          SE           28.0        SSE  ...        51.0         24.0   
145459         NaN            NaN        ESE  ...        62.0         36.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
145455       1024.6       1020.3       NaN       NaN     10.1     22.4   
145456       1023.5       1019.1       NaN       NaN     10.9     24.5   
145457       1021.0       1016.8       NaN       NaN     12.5     26.1   
145458       1019.4       1016.5       3.0       2.0     15.1     26.0   
145459       1020.2       1017.9       8.0       8.0     15.0     20.9   

        RainToday  RainTomorrow  
145455         No            No  
145456         No            No  
145457         No            No  
145458         No            No  
145459         No           NaN  

[5 rows x 23 columns]

with the df.tail() we can see that bottom 5 rows and 23 colums are there

In [ ]:
df.sample(5)

Date          Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
46822  3/20/2011          Canberra     14.0     21.7       0.6          2.4   
84036  7/30/2008          Brisbane      6.3     20.6       0.0          3.8   
6593   6/29/2010             Cobar      0.0     12.7       0.0          1.4   
65515  9/16/2012  MelbourneAirport      5.2     16.3       0.0          2.8   
13571  5/13/2013             Moree     13.4     23.5       0.0          4.0   

       Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
46822       2.9           E           44.0        SSE  ...        95.0   
84036       9.5         WNW           22.0        WSW  ...        55.0   
6593       10.6           N           24.0         NE  ...        67.0   
65515       0.7           N           26.0          N  ...        71.0   
13571       2.0         NNE           33.0        NNE  ...        59.0   

       Humidity3pm  Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  \
46822         68.0       1014.5       1012.0       8.0       6.0     15.7   
84036         25.0       1017.5       1014.2       0.0       0.0     13.8   
6593          39.0       1022.6       1019.7       0.0       0.0      6.6   
65515         51.0       1021.0       1018.3       7.0       7.0     11.5   
13571         44.0       1020.7       1015.5       7.0       7.0     18.3   

       Temp3pm  RainToday  RainTomorrow  
46822     19.6         No            No  
84036     20.3         No            No  
6593      11.8         No            No  
65515     15.3         No           Yes  
13571     22.4         No           Yes  

[5 rows x 23 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

    
    Based on the provided dataset, we can draw the following conclusions:
*   The dataset consists of 23 columns, with entries ranging from 0 to 86876. Among these columns, 16 are of the float64 data type, while 7 are of the object data type.
*   The dataset contains a total of 86877 entries but some of the columns occurs missing values
*   The dataset occupies 15.2 MB of RAM.







## **Missing Values**


In [ ]:
df.isnull()

Date  Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0       False     False    False    False     False         True      True   
1       False     False    False    False     False         True      True   
2       False     False    False    False     False         True      True   
3       False     False    False    False     False         True      True   
4       False     False    False    False     False         True      True   
...       ...       ...      ...      ...       ...          ...       ...   
145455  False     False    False    False     False         True      True   
145456  False     False    False    False     False         True      True   
145457  False     False    False    False     False         True      True   
145458  False     False    False    False     False         True      True   
145459  False     False    False     True     False         True      True   

        WindGustDir  WindGustSpeed  WindDir9am  ...  Humidity9am  Humidity3pm  \
0             False          False       False  ...        False        False   
1             False          False       False  ...        False        False   
2             False          False       False  ...        False        False   
3             False          False       False  ...        False        False   
4             False          False       False  ...        False        False   
...             ...            ...         ...  ...          ...          ...   
145455        False          False       False  ...        False        False   
145456        False          False       False  ...        False        False   
145457        False          False       False  ...        False        False   
145458        False          False       False  ...        False        False   
145459         True           True       False  ...        False        False   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0             False        False     False      True    False    False   
1             False        False      True      True    False    False   
2             False        False      True     False    False    False   
3             False        False      True      True    False    False   
4             False        False     False     False    False    False   
...             ...          ...       ...       ...      ...      ...   
145455        False        False      True      True    False    False   
145456        False        False      True      True    False    False   
145457        False        False      True      True    False    False   
145458        False        False     False     False    False    False   
145459        False        False     False     False    False    False   

        RainToday  RainTomorrow  
0           False         False  
1           False         False  
2           False         False  
3           False         False  
4           False         False  
...           ...           ...  
145455      False         False  
145456      False         False  
145457      False         False  
145458      False         False  
145459      False          True  

[145460 rows x 23 columns]

In [ ]:
df.isnull().sum()

Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64

**Here is a list of columns from the dataset along with the corresponding number of missing values:**

    MinTemp: 1196 missing values
    MaxTemp: 1012 missing values
    Rainfall: 2342 missing values
    Evaporation: 39660 missing values
    Sunshine: 46580 missing values
    WindGustDir: 6281 missing values
    WindGustSpeed: 6273 missing values
    WindDir9am: 7785 missing values
    WindDir3pm: 2970 missing values
    WindSpeed9am: 1448 missing values
    WindSpeed3pm: 2067 missing values
    Humidity9am: 2066 missing values
    Humidity3pm: 2487 missing values
    Pressure9am: 10656 missing values
    Pressure3pm: 10609 missing values
    Cloud9am: 33478 missing values
    Cloud3pm: 34502 missing values
    Temp9am: 1563 missing values
    Temp3pm: 2040 missing values
    RainToday: 2342 missing values
    RainTomorrow: 2340 missing values

In [ ]:
df.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [ ]:
df

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0       12/1/2008   Albury     13.4     22.9       0.6          NaN       NaN   
1       12/2/2008   Albury      7.4     25.1       0.0          NaN       NaN   
2       12/3/2008   Albury     12.9     25.7       0.0          NaN       NaN   
3       12/4/2008   Albury      9.2     28.0       0.0          NaN       NaN   
4       12/5/2008   Albury     17.5     32.3       1.0          NaN       NaN   
...           ...      ...      ...      ...       ...          ...       ...   
145455  6/21/2017    Uluru      2.8     23.4       0.0          NaN       NaN   
145456  6/22/2017    Uluru      3.6     25.3       0.0          NaN       NaN   
145457  6/23/2017    Uluru      5.4     26.9       0.0          NaN       NaN   
145458  6/24/2017    Uluru      7.8     27.0       0.0          NaN       NaN   
145459  6/25/2017    Uluru     14.9      NaN       0.0          NaN       NaN   

       WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0                W           44.0          W  ...        71.0         22.0   
1              WNW           44.0        NNW  ...        44.0         25.0   
2              WSW           46.0          W  ...        38.0         30.0   
3               NE           24.0         SE  ...        45.0         16.0   
4                W           41.0        ENE  ...        82.0         33.0   
...            ...            ...        ...  ...         ...          ...   
145455           E           31.0         SE  ...        51.0         24.0   
145456         NNW           22.0         SE  ...        56.0         21.0   
145457           N           37.0         SE  ...        53.0         24.0   
145458          SE           28.0        SSE  ...        51.0         24.0   
145459         NaN            NaN        ESE  ...        62.0         36.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
1            1010.6       1007.8       NaN       NaN     17.2     24.3   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
3            1017.6       1012.8       NaN       NaN     18.1     26.5   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
...             ...          ...       ...       ...      ...      ...   
145455       1024.6       1020.3       NaN       NaN     10.1     22.4   
145456       1023.5       1019.1       NaN       NaN     10.9     24.5   
145457       1021.0       1016.8       NaN       NaN     12.5     26.1   
145458       1019.4       1016.5       3.0       2.0     15.1     26.0   
145459       1020.2       1017.9       8.0       8.0     15.0     20.9   

        RainToday  RainTomorrow  
0              No            No  
1              No            No  
2              No            No  
3              No            No  
4              No            No  
...           ...           ...  
145455         No            No  
145456         No            No  
145457         No            No  
145458         No            No  
145459         No           NaN  

[145460 rows x 23 columns]

In [ ]:
df["Date"]

0         12/1/2008
1         12/2/2008
2         12/3/2008
3         12/4/2008
4         12/5/2008
            ...    
145455    6/21/2017
145456    6/22/2017
145457    6/23/2017
145458    6/24/2017
145459    6/25/2017
Name: Date, Length: 145460, dtype: object

In [ ]:
print(df['Date'].min())
print(df['Date'].max())

1/1/2008
9/9/2016


Dataset of starting date is 01-11-2007 and ending date is 25-06-2017

In [ ]:
df.T

0          1          2          3          4       \
Date           12/1/2008  12/2/2008  12/3/2008  12/4/2008  12/5/2008   
Location          Albury     Albury     Albury     Albury     Albury   
MinTemp             13.4        7.4       12.9        9.2       17.5   
MaxTemp             22.9       25.1       25.7       28.0       32.3   
Rainfall             0.6        0.0        0.0        0.0        1.0   
Evaporation          NaN        NaN        NaN        NaN        NaN   
Sunshine             NaN        NaN        NaN        NaN        NaN   
WindGustDir            W        WNW        WSW         NE          W   
WindGustSpeed       44.0       44.0       46.0       24.0       41.0   
WindDir9am             W        NNW          W         SE        ENE   
WindDir3pm           WNW        WSW        WSW          E         NW   
WindSpeed9am        20.0        4.0       19.0       11.0        7.0   
WindSpeed3pm        24.0       22.0       26.0        9.0       20.0   
Humidity9am         71.0       44.0       38.0       45.0       82.0   
Humidity3pm         22.0       25.0       30.0       16.0       33.0   
Pressure9am       1007.7     1010.6     1007.6     1017.6     1010.8   
Pressure3pm       1007.1     1007.8     1008.7     1012.8     1006.0   
Cloud9am             8.0        NaN        NaN        NaN        7.0   
Cloud3pm             NaN        NaN        2.0        NaN        8.0   
Temp9am             16.9       17.2       21.0       18.1       17.8   
Temp3pm             21.8       24.3       23.2       26.5       29.7   
RainToday             No         No         No         No         No   
RainTomorrow          No         No         No         No         No   

                  5          6          7          8           9       ...  \
Date           12/6/2008  12/7/2008  12/8/2008  12/9/2008  12/10/2008  ...   
Location          Albury     Albury     Albury     Albury      Albury  ...   
MinTemp             14.6       14.3        7.7        9.7        13.1  ...   
MaxTemp             29.7       25.0       26.7       31.9        30.1  ...   
Rainfall             0.2        0.0        0.0        0.0         1.4  ...   
Evaporation          NaN        NaN        NaN        NaN         NaN  ...   
Sunshine             NaN        NaN        NaN        NaN         NaN  ...   
WindGustDir          WNW          W          W        NNW           W  ...   
WindGustSpeed       56.0       50.0       35.0       80.0        28.0  ...   
WindDir9am             W         SW        SSE         SE           S  ...   
WindDir3pm             W          W          W         NW         SSE  ...   
WindSpeed9am        19.0       20.0        6.0        7.0        15.0  ...   
WindSpeed3pm        24.0       24.0       17.0       28.0        11.0  ...   
Humidity9am         55.0       49.0       48.0       42.0        58.0  ...   
Humidity3pm         23.0       19.0       19.0        9.0        27.0  ...   
Pressure9am       1009.2     1009.6     1013.4     1008.9      1007.0  ...   
Pressure3pm       1005.4     1008.2     1010.1     1003.6      1005.7  ...   
Cloud9am             NaN        1.0        NaN        NaN         NaN  ...   
Cloud3pm             NaN        NaN        NaN        NaN         NaN  ...   
Temp9am             20.6       18.1       16.3       18.3        20.1  ...   
Temp3pm             28.9       24.6       25.5       30.2        28.2  ...   
RainToday             No         No         No         No         Yes  ...   
RainTomorrow          No         No         No        Yes          No  ...   

                  145450     145451     145452     145453     145454  \
Date           6/16/2017  6/17/2017  6/18/2017  6/19/2017  6/20/2017   
Location           Uluru      Uluru      Uluru      Uluru      Uluru   
MinTemp              5.2        6.4        8.0        7.4        3.5   
MaxTemp             24.3       23.4       20.7       20.6       21.8   
Rainfall             0.0        0.0        0.0        0.

In [ ]:
df.iloc[::2]

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0       12/1/2008   Albury     13.4     22.9       0.6          NaN       NaN   
2       12/3/2008   Albury     12.9     25.7       0.0          NaN       NaN   
4       12/5/2008   Albury     17.5     32.3       1.0          NaN       NaN   
6       12/7/2008   Albury     14.3     25.0       0.0          NaN       NaN   
8       12/9/2008   Albury      9.7     31.9       0.0          NaN       NaN   
...           ...      ...      ...      ...       ...          ...       ...   
145450  6/16/2017    Uluru      5.2     24.3       0.0          NaN       NaN   
145452  6/18/2017    Uluru      8.0     20.7       0.0          NaN       NaN   
145454  6/20/2017    Uluru      3.5     21.8       0.0          NaN       NaN   
145456  6/22/2017    Uluru      3.6     25.3       0.0          NaN       NaN   
145458  6/24/2017    Uluru      7.8     27.0       0.0          NaN       NaN   

       WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0                W           44.0          W  ...        71.0         22.0   
2              WSW           46.0          W  ...        38.0         30.0   
4                W           41.0        ENE  ...        82.0         33.0   
6                W           50.0         SW  ...        49.0         19.0   
8              NNW           80.0         SE  ...        42.0          9.0   
...            ...            ...        ...  ...         ...          ...   
145450           E           24.0         SE  ...        53.0         24.0   
145452         ESE           41.0         SE  ...        56.0         32.0   
145454           E           31.0        ESE  ...        59.0         27.0   
145456         NNW           22.0         SE  ...        56.0         21.0   
145458          SE           28.0        SSE  ...        51.0         24.0   

        Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
0            1007.7       1007.1       8.0       NaN     16.9     21.8   
2            1007.6       1008.7       NaN       2.0     21.0     23.2   
4            1010.8       1006.0       7.0       8.0     17.8     29.7   
6            1009.6       1008.2       1.0       NaN     18.1     24.6   
8            1008.9       1003.6       NaN       NaN     18.3     30.2   
...             ...          ...       ...       ...      ...      ...   
145450       1023.8       1020.0       NaN       NaN     12.3     23.3   
145452       1028.1       1024.3       NaN       7.0     11.6     20.0   
145454       1024.7       1021.2       NaN       NaN      9.4     20.9   
145456       1023.5       1019.1       NaN       NaN     10.9     24.5   
145458       1019.4       1016.5       3.0       2.0     15.1     26.0   

        RainToday  RainTomorrow  
0              No            No  
2              No            No  
4              No            No  
6              No            No  
8              No           Yes  
...           ...           ...  
145450         No            No  
145452         No            No  
145454         No            No  
145456         No            No  
145458         No            No  

[72730 rows x 23 columns]

### What is the shape of given dataset ?

In [ ]:
df.shape


(145460, 23)

In [ ]:
df.size

3345580

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

How many integer columns are given in the dataset?


In [ ]:
df.dtypes


Date              object
Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Cloud9am         float64
Cloud3pm         float64
Temp9am          float64
Temp3pm          float64
RainToday         object
RainTomorrow      object
dtype: object

In [ ]:
df.dtypes.value_counts()

float64    16
object      7
dtype: int64

What is the datatype of ' WindGustSpeed' column?
*

In [ ]:
df.WindGustSpeed

0         44.0
1         44.0
2         46.0
3         24.0
4         41.0
          ... 
145455    31.0
145456    22.0
145457    37.0
145458    28.0
145459     NaN
Name: WindGustSpeed, Length: 145460, dtype: float64

In [ ]:
df.WindGustSpeed.dtypes

dtype('float64')

How many missing values are there in 'Sunshine' column?
*

In [ ]:
df.Sunshine

0        NaN
1        NaN
2        NaN
3        NaN
4        NaN
          ..
145455   NaN
145456   NaN
145457   NaN
145458   NaN
145459   NaN
Name: Sunshine, Length: 145460, dtype: float64

In [ ]:
df.Sunshine.dtypes

dtype('float64')

In [ ]:
df['Sunshine'].isnull()

0         True
1         True
2         True
3         True
4         True
          ... 
145455    True
145456    True
145457    True
145458    True
145459    True
Name: Sunshine, Length: 145460, dtype: bool

In [ ]:
df['Sunshine'].isnull().sum()

69835

In [ ]:
df.Sunshine.isnull().sum()-  18441

51394

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

How many non-null values there in ' Humidity9am' column?
*

In [ ]:
df.Humidity9am

0         71.0
1         44.0
2         38.0
3         45.0
4         82.0
          ... 
145455    51.0
145456    56.0
145457    53.0
145458    51.0
145459    62.0
Name: Humidity9am, Length: 145460, dtype: float64

In [ ]:
df['Humidity9am'].isnull().sum()

2654

In [ ]:
df['Humidity9am'].count()

142806

In [ ]:
df['Humidity9am'].count()-df['Humidity9am'].isnull().sum()

140152

How many object columns are there in the given dataset?
*

In [ ]:
df.dtypes

Date              object
Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Cloud9am         float64
Cloud3pm         float64
Temp9am          float64
Temp3pm          float64
RainToday         object
RainTomorrow      object
dtype: object

In [ ]:
df.dtypes.value_counts()   # there are 7 oject

float64    16
object      7
dtype: int64

How much memory is occupied by the dataset on your computers RAM?



In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

**memory usage: 9.5+ MB**

Less than 25 MB

What is the percentage of values missing in 'MinTemp' column?
*

In [ ]:
df['MinTemp'].isnull().sum()

1485

In [ ]:
df['MinTemp'].isnull().sum()*100

148500

In [ ]:
percent_missing = df.isnull().sum() * 100 / len(df)
missing_value_df = pd.DataFrame({'column_name': df.columns,
                                 'percent_missing': percent_missing})

In [ ]:
percentage_missing_values=df.isnull().sum()*100/len(df)

In [ ]:
df['MinTemp'].isnull().sum()*100/len(df)

1.0208992162793895

Did it rain in 'Uluru' location on '23rd June 2017' ?


In [ ]:

df['Rainfall']

0         0.6
1         0.0
2         0.0
3         0.0
4         1.0
         ... 
145455    0.0
145456    0.0
145457    0.0
145458    0.0
145459    0.0
Name: Rainfall, Length: 145460, dtype: float64

In [ ]:
uluru_rain = df[(df['Location'] == 'Uluru') & (df['Date'] == '2017-06-23')]
uluru_rain


Empty DataFrame
Columns: [Date, Location, MinTemp, MaxTemp, Rainfall, Evaporation, Sunshine, WindGustDir, WindGustSpeed, WindDir9am, WindDir3pm, WindSpeed9am, WindSpeed3pm, Humidity9am, Humidity3pm, Pressure9am, Pressure3pm, Cloud9am, Cloud3pm, Temp9am, Temp3pm, RainToday, RainTomorrow]
Index: []

[0 rows x 23 columns]

What was the temperature at 3 PM in  'Albury' location on '3rd Dec 2008' ?


Which of the following is correct?
*

Out of total 145460 days, on how many days it rained?
*

In [ ]:
df['Rainfall'].value_counts()

0.0      91080
0.2       8761
0.4       3782
0.6       2592
0.8       2056
         ...  
134.8        1
84.4         1
157.6        1
166.8        1
69.0         1
Name: Rainfall, Length: 681, dtype: int64

Variance in rainfall recorded is?


In [ ]:
df.var()

<ipython-input-103-28ded241fd7c>:1: FutureWarning: The default value of numeric_only in DataFrame.var is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.var()


MinTemp           40.940738
MaxTemp           50.680856
Rainfall          71.877497
Evaporation       17.587154
Sunshine          14.329881
WindGustSpeed    185.152144
WindSpeed9am      79.483917
WindSpeed3pm      77.612576
Humidity9am      362.109100
Humidity3pm      432.469526
Pressure9am       50.502773
Pressure3pm       49.525193
Cloud9am           8.335686
Cloud3pm           7.400344
Temp9am           42.103917
Temp3pm           48.117120
dtype: float64

In [ ]:
df['Rainfall'].var()

71.87749691645696

Which Australian Town/City data is not available in the given dataset?


In [ ]:
df['Location'].values

array(['Albury', 'Albury', 'Albury', ..., 'Uluru', 'Uluru', 'Uluru'],
      dtype=object)

In [ ]:
# Which Australian Town/City data is not available in the given dataset?

col = ['GoldCoast', 'Albany', 'Perth', 'Orange', 'Melbourne']

for location in col:
    if location not in df['Location'].values:
        print(location)



Orange


Choose all the correct options:

    Minimum MaxTemp observed is -4.80
    25th Percentile of Wind Speed at 9 AM is 7.00
    Median Temperature observed at 3 PM was 21.10
    On average 5.468232 evaporation is recorded.
    None of the above

In [ ]:
df['MaxTemp'].min()

-4.8

In [ ]:
per_25= np.percentile(df['WindSpeed9am'], 25)
print(per_25)

nan


In [ ]:
df['Temp3pm'].median()

21.1

In [ ]:
df['Evaporation'].mean()

5.468231522922462

Out of total 145460 days, on how many days it rained?


In [ ]:

rainy_days=df[df['RainToday']=="Yes"]['RainToday'].count()
print(rainy_days)

31880
